### ISSUE TITLE
Deploy react-static as normal SPA

### ISSUE THREAD

**Issue by simonunlikely** [nozzle/react-static#1173]
Hey guys,

I wonder if it's possible to deploy a react-static app as a SPA (pulling dynamic data in browser without actually building html) so I can host a "preview" app where contributors would be able to see how changes look _before_ triggering a full website rebuild that can take quite some time depending on the website.

Don't know if that makes sense or if there is a better way to do this ?

Thanks !

**Comment by krzysztof-pyrzyk-ah** [comment]
Hey,
It is possible.
Have a look here: https://github.com/pauloamgomes/CockpitCMS-React-Static

**Comment by simonunlikely** [comment]
Hey, thanks for your fast answer :)

This is an interesting example on how to achieve this ! However in this example the preview lives under a special page, which is something I cannot do as I have multiple page templates that are really different. I would also like to fetch my data instead of it being pushed from a postMessage event

I wonder if it would be possible to have my front end parse any url having the param preview=true and then rerender using preview data and not builtin data ? What would be great is being able to reload the getRoutes() from the browser. I see how I could test for the preview param and then trigger a manual fetch to get last data and update my page, but it would just be simpler if I could use what i've already written in my static.config...

Don't know if that makes sense ?

**Comment by sw-yx** [comment]
the answer is yes, but there are some things you wont be able to do. for example in static.config.js you have access to the full power of Node.js. however if you want to do that within the browser for your SPA you wont have this ability (unless you stick it in a serverless function or something...)



as an alternative, if you deploy thru netlify, every PR will get a deploy preview: https://www.netlify.com/blog/2016/07/20/introducing-deploy-previews-in-netlify/ that may help.

**Comment by simonunlikely** [comment]
Thanks for your answer! That makes sense. I wonder if there would be a way to share route information between static.config.js and the browser ? Like having a seperate file specifying data dependencies for each route that I can then import in static.config.js for static rendering, and that I can also import directly in my dynamic preview component so I know what data to fetch for what template?

**Comment by Pixelatex** [comment]
@simonunlikely So, I did something like this once but it's not a quick fix.

I had a file to contains data references and where to get them, e.g. posts come from 'api/posts' and such.

My static.config imports these and runs them during build, the usual stuff.

Now on to how I made it work, I wrap each page/template in a dynamic-data-hoc. If the hoc detects `?preview=true` in the url, it will start fetching data.

To know what data I need, I look into the keys provided to the props of my template and see if any of them match data references, if so, fetch them and replace the old props with these props.

Your component then receives 'new' data instead of the static data, I did have to add a small loader to show that it was refetching the new data and such.

This way by far the 'easiest' way I could figure out how to preview pages without running your site as a separate SPA  on a preview server.
